File: app/src/main/java/com/karkoszka/cookingtime/activities/MainActivity.java
package com.karkoszka.cookingtime.activities;

import android.app.AlarmManager;
import android.app.KeyguardManager;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.PowerManager;
import android.os.SystemClock;
import androidx.core.app.NotificationCompat;
import androidx.core.app.TaskStackBuilder;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Chronometer;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.karkoszka.cookingtime.R;
import com.karkoszka.cookingtime.common.LoaderPreferences;
import com.karkoszka.cookingtime.common.Plate;
import com.karkoszka.cookingtime.fragments.MainScreen6Fragment;
import com.karkoszka.cookingtime.services.AlarmSoundService;
import com.karkoszka.cookingtime.services.CTBroadcastReceiver;


public class MainActivity extends AppCompatActivity implements
	MainScreen6Fragment.OnMainScreenFragmentInteractionListener {
	
	public static final String ALARM_OFF_PLATE_NO = "Alarm_Off_Plate_No";
	private static final int NOTIFICATION_ID = 486534515;
	
	private static final int ALARM_UNIQUE_PREFIX = 68923402;
	private LoaderPreferences loader;
	private Plate[] plates = new Plate[6];
	private TextView[] plateAIT = new TextView[6];
	private Chronometer[] chronos = new Chronometer[6];
	private ImageButton[] startButtons = new ImageButton[6];
	private FrameLayout[] colorFrame = new FrameLayout[6];
	private PendingIntent[] pIntents = new PendingIntent[6]; 
	private BroadcastReceiver receiver = new CTBroadcastReceiver();
	private boolean alarmSoundBlockSet = false;
	
    @Override
    public void onCreate(Bundle savedInstanceState) {
        Log.d("On create", "" + SystemClock.elapsedRealtime());
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        registerReceiver(receiver, new IntentFilter());
        SharedPreferences settings =
        		getSharedPreferences(LoaderPreferences.PREFS_NAME, 0);
        loader = new LoaderPreferences(settings);
        initUIControls();

		PowerManager pm = (PowerManager) getApplicationContext().getSystemService(Context.POWER_SERVICE);
		PowerManager.WakeLock wakeLock = pm.newWakeLock((PowerManager.SCREEN_BRIGHT_WAKE_LOCK | PowerManager.FULL_WAKE_LOCK | PowerManager.ACQUIRE_CAUSES_WAKEUP), "com.karkoszka.cookingtime:waketag");
		wakeLock.acquire();

		KeyguardManager keyguardManager = (KeyguardManager) getApplicationContext().getSystemService(Context.KEYGUARD_SERVICE);
		KeyguardManager.KeyguardLock keyguardLock = keyguardManager.newKeyguardLock("com.karkoszka.cookingtime:waketag");
		keyguardLock.disableKeyguard();
    }
	@Override
    public void onResume() {
        Log.d("On Resume", "" + SystemClock.elapsedRealtime());
    	super.onResume();
        plates = loader.loadPlates();
        Intent intent = getIntent();
        if(intent.hasExtra(ALARM_OFF_PLATE_NO)) {
        	int pl = getIntent().getIntExtra(ALARM_OFF_PLATE_NO, 100);
        	if(plates[pl].getRuns() == Plate.STARTED && !alarmSoundBlockSet) {
        		alarmSoundBlockSet = true;
        		alarmOffToast(pl);
        	}
        }
		for(int i = 0;i < 6;i++) {
			makeAlarmInfoText(i);
        	onResumeChronometers(i);
        	setBackground(i);
			highlightAlarms(i);
			plateChangedReorderAlarm(i);
		}
	}

	private void fireAlarm(int pl) {
		Toast.makeText(this,getResources().getString(R.string.alarm) + (pl + 1) + getResources().getString(R.string.passed), Toast.LENGTH_LONG)
				.show();
		Log.d("MA reconfiguring", "canceled: " + pl );
		Intent intentSoundService = new Intent(getApplicationContext(), AlarmSoundService.class);
		intentSoundService.putExtra(MainActivity.ALARM_OFF_PLATE_NO, pl);
		intentSoundService.putExtra(AlarmSoundService.START_PLAY, true);

		Log.i("SimpleWakefulReceiver", "Starting service @ " + SystemClock.elapsedRealtime());
		startService(intentSoundService);
	}

	private void plateChangedReorderAlarm(int i){
		if(plates[i].getRuns() == Plate.CHANGED) {
			plates[i].setRuns(Plate.STARTED);
			loader.savePlate(plates[i]);
			cancelAlarm(i);
			if(plates[i].checkIfFired())
				fireAlarm(i);
			else
				startAlarmFromBefore(i);
		}
	}
    /**
     * Makes info toast that plate is running
     * @param plate plates id
     */
    private void alarmOffToast(int plate) {
    	//int pl = plate + 1;
	    CharSequence msg = getResources().getString(R.string.timer) + (plate + 1) + getResources().getString(R.string.is_done);
	    Toast.makeText(this, msg, Toast.LENGTH_LONG).show();
	}
    /**
     * shows notification after push to start button
     */
    private void notificate() {
    	NotificationCompat.Builder mBuilder =
    	        new NotificationCompat.Builder(this)
    	        .setSmallIcon(R.drawable.ic_stat_six_timers_bw2)
    	        .setContentTitle( getResources().getString(R.string.running))
    	        .setAutoCancel(false);
    	Intent resultIntent = new Intent(this, MainActivity.class);
    	resultIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP|
				Intent.FLAG_ACTIVITY_SINGLE_TOP);
    	TaskStackBuilder stackBuilder = TaskStackBuilder.create(this);
    	stackBuilder.addParentStack(MainActivity.class);
    	stackBuilder.addNextIntent(resultIntent);
    	PendingIntent resultPendingIntent =
    	        stackBuilder.getPendingIntent(
    	            0,
    	            PendingIntent.FLAG_CANCEL_CURRENT
    	        );
    	mBuilder.setContentIntent(resultPendingIntent);
    	NotificationManager mNotificationManager =
    	    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
    	mNotificationManager.notify(NOTIFICATION_ID, mBuilder.build());
    }

    private void cancelNotification() {
    	NotificationManager mNotificationManager =
        	    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
    	mNotificationManager.cancelAll();
    }
    /**
	 * Stops alarms service
	 */
	private void alarmSoundServiceStop() {
		Intent intent = new Intent(getApplicationContext(), AlarmSoundService.class);
		stopService(intent);
		alarmSoundBlockSet = false;
	}
	private boolean isAlarm() {
		for(Plate pl : plates) {
			if(pl.getRuns() == 1)
				return true;
		}
		return false;
	}

	private void highlightAlarms(int i) {
		if(plates[i].getRuns() == Plate.STARTED && plates[i].checkIfFired())
    		plateAIT[i].setText("\\\\\\o  " + plateAIT[i].getText() + "  o///");
    }

	public void setPlate1(View view) {
		this.clickSetButton(0);
	}
	public void setPlate2(View view) {
		this.clickSetButton(1);
	}
	public void setPlate3(View view) {
		this.clickSetButton(2);
	}
	public void setPlate4(View view) {
		this.clickSetButton(3);
	}
	public void setPlate5(View view) {
		this.clickSetButton(4);
	}
	public void setPlate6(View view) {
		this.clickSetButton(5);
	}

	private void clickSetButton(int plate) {
		if (!alarmSoundBlockSet) {
			Intent intent = new Intent(this, SetPlateActivity.class);
			intent.putExtra(SetPlateActivity.PLATE, plate);
			startActivity(intent);
		}
	}

	public void startPlate1(View view) {
		this.clickStartButton(0);
	}
	public void startPlate2(View view) {
		this.clickStartButton(1);
	}
	public void startPlate3(View view) {
		this.clickStartButton(2);
	}
	public void startPlate4(View view) {
		this.clickStartButton(3);
	}
	public void startPlate5(View view) {
		this.clickStartButton(4);
	}
	public void startPlate6(View view) {
		this.clickStartButton(5);
	}
	/**
	 * Serve start buttons and chronometers according plates state
	 */
	private void clickStartButton(int plate) {
		Chronometer chrono = this.chronos[plate];
		ImageButton button = this.startButtons[plate];

		if(this.plates[plate].isReady() && this.plates[plate].computeSetOff() > 0){
	        startAlarmFromNow(plate);
			chrono.setBase(this.plates[plate].getBaseForChronometer());
			chrono.start();
			button.setImageDrawable(getResources().getDrawable(R.drawable.outline_stop_black_36));
			notificate();
    	}

		else if (this.plates[plate].isStarted()) {
			stopAlarm(plate);
			for(Plate alarmed: plates) {
				if(alarmed.isStarted() && alarmed.checkIfFired())
					stopAlarm(alarmed.getId());
			}

    	}
		else if (this.plates[plate].isStopped()) {
			chrono.setBase(SystemClock.elapsedRealtime());
            button.setImageDrawable(getResources().getDrawable(R.drawable.outline_play_arrow_black_36));
    		this.plates[plate].reset();
			loader.savePlate(plate, plates[plate].getRuns(), "");
    	}
	}

	private void stopAlarm(int plate) {
		this.chronos[plate].stop();
		cancelAlarm(plate);
		makeAlarmInfoText(plate);
		alarmSoundServiceStop();
        startButtons[plate].setImageDrawable(getResources().getDrawable(R.drawable.baseline_replay_black_36));
		this.plates[plate].stop();
		if(isAlarm())
			notificate();
		else
			cancelNotification();
		this.plates[plate].setLast(this.chronos[plate].getText().toString());
		loader.savePlate(plate, plates[plate].getRuns(), plates[plate].getLast());
	}

	private void startAlarmFromNow(int plate) {
        startIntentAndManager(plate, plates[plate].startFromNow());
	}

	private void startAlarmFromBefore(int plate) {
        startIntentAndManager(plate, plates[plate].startFromBefore());
    }

	private void startIntentAndManager(int plate, long time) {
        Intent intent = new Intent(this, CTBroadcastReceiver.class);
        intent.putExtra(ALARM_OFF_PLATE_NO, plate);
        pIntents[plate] = PendingIntent.getBroadcast(this.getApplicationContext()
                , ALARM_UNIQUE_PREFIX + plate
                , intent
                , PendingIntent.FLAG_CANCEL_CURRENT);
        ((AlarmManager) this.getSystemService(Context.ALARM_SERVICE))
                .setExact(AlarmManager.RTC_WAKEUP
                        , time
                        , pIntents[plate]);
        loader.savePlate(plate, plates[plate].getBase(),plates[plate].getSetOff()
                ,plates[plate].getRuns());
    }

    private void cancelAlarm(int plate) {
		Log.d("MA reconfiguring ", "plateR from ST: " + plate);
		if(pIntents[plate] != null) {
			((AlarmManager) this.getSystemService(Context.ALARM_SERVICE))
			  .cancel(pIntents[plate]);
			pIntents[plate] = null;
		}
		else {
			((AlarmManager) this.getSystemService(Context.ALARM_SERVICE))
			  .cancel(PendingIntent.getBroadcast(this.getApplicationContext(),
						ALARM_UNIQUE_PREFIX + plate,
						(new Intent(this,CTBroadcastReceiver.class))
							.putExtra(ALARM_OFF_PLATE_NO, plate),
							PendingIntent.FLAG_CANCEL_CURRENT));
		}
	}

	private void onResumeChronometers(int i) {
			if (plates[i].getRuns() == Plate.STARTED)
				plateIsStartedOnResume(i);
			else if (plates[i].getRuns() == Plate.STOPPED)
				plateIsStoppedOnResume(i);
			else if (plates[i].getRuns() == Plate.READY)
				plateIsReadyOnResume(i);
	}

	private void plateIsReadyOnResume(int i) {
		chronos[i].setBase(SystemClock.elapsedRealtime());
		startButtons[i].setImageDrawable(getResources().getDrawable(R.drawable.outline_play_arrow_black_36));
	}

	private void plateIsStoppedOnResume(int i) {
		chronos[i].setText(plates[i].getLast());
		startButtons[i].setImageDrawable(getResources().getDrawable(R.drawable.baseline_replay_black_36));
	}

	private void plateIsStartedOnResume(int i) {
		chronos[i].setBase(plates[i].getBaseForChronometer());
		Log.d("Chronometer base: ", "" + plates[i].getBase());
		chronos[i].start();
		startButtons[i].setImageDrawable(getResources().getDrawable(R.drawable.outline_stop_black_36));
	}

	public void onBroadcastReceived(int plate) {
		this.plateAIT[plate].setText("\\\\\\o" + plates[plate].getHours() + ":"
				+ plates[plate].getMinutes() + ":" + plates[plate].getSeconds());
	}

	@Override
	protected void onDestroy() {
	       super.onDestroy();
	       unregisterReceiver(receiver);
	}

	@Override
	public void onStartPressed(int plate) {
        //default implementation ignored
	}


	@Override
	public void onSetPressed(int plate) {
	    //default implementation ignored
	}

    /**
     * sets one plates alarm info text to normal 
     */
    private void makeAlarmInfoText(int plate) {
    	if (plates[plate].getHours() != 0) {
			this.plateAIT[plate].setText("" + formateToTwoDigits(plates[plate].getHours())
					+ ":"
					+ formateToTwoDigits(plates[plate].getMinutes()) 
					+ ":" 
					+ formateToTwoDigits(plates[plate].getSeconds()));
		} else {
			this.plateAIT[plate].setText("" + formateToTwoDigits(plates[plate].getMinutes()) 
					+ ":" 
					+ formateToTwoDigits(plates[plate].getSeconds()));
		}
    }

    /**
     * Converts single time info to two digit formate 
     * @param num time cell digit in int 
     * @return 
     */
    private String formateToTwoDigits(int num) {
    	if(num < 10) {
    		return "0" + Integer.toString(num);
    	}
    	return Integer.toString(num);
    }

    private void setBackground(int i) {
    		colorFrame[i].setBackgroundColor(plates[i].getColour());
    }

    /**
     * inits UI controls
     */
    private void initUIControls() {
    	initAlarmInfoText();
        initChronometers();
        initStartButtons();
        initBackground();
	}

	/**
     * Obtains references of alarm info texts
     */
    private void initAlarmInfoText() {
    	this.plateAIT[0] = (TextView) findViewById(R.id.timeInfo1);
    	this.plateAIT[1] = (TextView) findViewById(R.id.timeInfo2);
    	this.plateAIT[2] = (TextView) findViewById(R.id.timeInfo3);
    	this.plateAIT[3] = (TextView) findViewById(R.id.timeInfo4);
    	this.plateAIT[4] = (TextView) findViewById(R.id.timeInfo5);
    	this.plateAIT[5] = (TextView) findViewById(R.id.timeInfo6);
    }

    /**
     * Obtains references of chronometers
     */
    private void initChronometers() {
    	this.chronos[0] = (Chronometer) findViewById(R.id.chronometer1);
    	this.chronos[1] = (Chronometer) findViewById(R.id.chronometer2);
    	this.chronos[2] = (Chronometer) findViewById(R.id.chronometer3);
    	this.chronos[3] = (Chronometer) findViewById(R.id.chronometer4);
    	this.chronos[4] = (Chronometer) findViewById(R.id.chronometer5);
    	this.chronos[5] = (Chronometer) findViewById(R.id.chronometer6);
    }

    /**
     * Obtains references of start buttons
     */
    private void initStartButtons() {
    	this.startButtons[0] = (ImageButton) findViewById(R.id.button1start);
    	this.startButtons[1] = (ImageButton) findViewById(R.id.button2start);
    	this.startButtons[2] = (ImageButton) findViewById(R.id.button3start);
    	this.startButtons[3] = (ImageButton) findViewById(R.id.button4start);
    	this.startButtons[4] = (ImageButton) findViewById(R.id.button5start);
    	this.startButtons[5] = (ImageButton) findViewById(R.id.button6start);
    }

    /**
     * Obtains references of background frames
     */
    private void initBackground() {
    	colorFrame[0] = (FrameLayout) findViewById(R.id.frame1);
    	colorFrame[1] = (FrameLayout) findViewById(R.id.frame2);
    	colorFrame[2] = (FrameLayout) findViewById(R.id.frame3);
    	colorFrame[3] = (FrameLayout) findViewById(R.id.frame4);
    	colorFrame[4] = (FrameLayout) findViewById(R.id.frame5);
    	colorFrame[5] = (FrameLayout) findViewById(R.id.frame6);
    }
}

File: app/src/main/java/com/karkoszka/cookingtime/activities/MainActivity.kt
package com.karkoszka.cookingtime.activities

import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.os.Bundle
import android.os.PowerManager
import android.os.SystemClock
import android.view.View
import android.view.WindowManager
import android.widget.Chronometer
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder
import androidx.core.content.res.ResourcesCompat
import com.karkoszka.cookingtime.R
import com.karkoszka.cookingtime.common.ColorFrame
import com.karkoszka.cookingtime.common.IColorFrame
import com.karkoszka.cookingtime.common.LoaderPreferences
import com.karkoszka.cookingtime.common.Plate
import com.karkoszka.cookingtime.fragments.MainScreen6Fragment.OnMainScreenFragmentInteractionListener
import com.karkoszka.cookingtime.services.AlarmSoundService
import com.karkoszka.cookingtime.services.CTBroadcastReceiver

class MainActivity : AppCompatActivity(), OnMainScreenFragmentInteractionListener {
    private var loader: LoaderPreferences? = null
    private var plates = arrayOfNulls<Plate>(6)
    private val plateAIT = arrayOfNulls<TextView>(6)
    private val chronos = arrayOfNulls<Chronometer>(6)
    private val startButtons = arrayOfNulls<ImageButton>(6)
    private val colorFrame = arrayOfNulls<IColorFrame>(6)
    private val pIntents = arrayOfNulls<PendingIntent>(6)
    private val receiver: BroadcastReceiver = CTBroadcastReceiver()
    private var alarmSoundBlockSet = false

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        registerReceiver(receiver, IntentFilter())
        val settings = getSharedPreferences(LoaderPreferences.PREFS_NAME, 0)
        loader = LoaderPreferences(settings)
        initUIControls()
        val wakeLock: PowerManager.WakeLock =
            (getSystemService(Context.POWER_SERVICE) as PowerManager).run {
                newWakeLock(
                    PowerManager.PARTIAL_WAKE_LOCK,
                    resources.getString(R.string.app_waketag)
                ).apply {
                    acquire()
                    release()
                }
            }
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    public override fun onResume() {
        super.onResume()
        plates = loader!!.loadPlates()
        val intent = intent
        if (intent.hasExtra(ALARM_OFF_PLATE_NO)) {
            val pl = getIntent().getIntExtra(ALARM_OFF_PLATE_NO, 100)
            if (plates[pl]!!.runs == Plate.STARTED && !alarmSoundBlockSet) {
                alarmSoundBlockSet = true
                alarmOffToast(pl)
            }
        }
        for (i in 0..5) {
            makeAlarmInfoText(i)
            onResumeChronometers(i)
            setBackground(i)
            highlightAlarms(i)
            plateChangedReorderAlarm(i)
        }
    }

    private fun fireAlarm(pl: Int) {
        Toast.makeText(
            this,
            String.format(resources.getString(R.string.alarm_off), (pl + 1)),
            Toast.LENGTH_LONG
        )
            .show()
        val intentSoundService = Intent(applicationContext, AlarmSoundService::class.java)
        intentSoundService.putExtra(ALARM_OFF_PLATE_NO, pl)
        intentSoundService.putExtra(AlarmSoundService.START_PLAY, true)
        startService(intentSoundService)
    }

    private fun plateChangedReorderAlarm(i: Int) {
        if (plates[i]!!.runs == Plate.CHANGED) {
            plates[i]!!.runs = Plate.STARTED
            loader!!.savePlate(plates[i]!!)
            cancelAlarm(i)
            if (plates[i]!!.checkIfFired()) fireAlarm(i) else startAlarmFromBefore(i)
        }
    }

    /**
     * Makes info toast that plate is running
     * @param plate plates id
     */
    private fun alarmOffToast(plate: Int) {
        val msg: CharSequence = String.format(resources.getString(R.string.timer_done), (plate + 1))
        Toast.makeText(this, msg, Toast.LENGTH_LONG).show()
    }

    /**
     * shows notification after push to start button
     */
    private fun notification() {
        val mNotificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create the NotificationChannel
            val mChannel = NotificationChannel(
                getString(R.string.channel_id),
                getString(R.string.channel_name),
                NotificationManager.IMPORTANCE_DEFAULT
            )
            val descriptionText = getString(R.string.channel_description)
            mChannel.description = descriptionText
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
//            val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            mNotificationManager.createNotificationChannel(mChannel)
        }
        val mBuilder = NotificationCompat.Builder(this, resources.getString(R.string.app_package))
            .setSmallIcon(R.drawable.ic_stat_six_timers_bw2)
            .setContentTitle(resources.getString(R.string.running))
            .setAutoCancel(false)
        val resultIntent = Intent(this, MainActivity::class.java)
        resultIntent.addFlags(
            Intent.FLAG_ACTIVITY_CLEAR_TOP or
                    Intent.FLAG_ACTIVITY_SINGLE_TOP
        )
        val stackBuilder = TaskStackBuilder.create(this)
        stackBuilder.addParentStack(MainActivity::class.java)
        stackBuilder.addNextIntent(resultIntent)
        val resultPendingIntent = stackBuilder.getPendingIntent(
            0,
            PendingIntent.FLAG_CANCEL_CURRENT
        )
        mBuilder.setContentIntent(resultPendingIntent)
        mNotificationManager.notify(NOTIFICATION_ID, mBuilder.build())
    }

    private fun cancelNotification() {
        val mNotificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        mNotificationManager.cancelAll()
    }

    /**
     * Stops alarms service
     */
    private fun alarmSoundServiceStop() {
        val intent = Intent(applicationContext, AlarmSoundService::class.java)
        stopService(intent)
        alarmSoundBlockSet = false
    }

    private val isAlarm: Boolean
        get() {
            for (pl in plates) {
                if (pl!!.runs == 1) return true
            }
            return false
        }

    private fun highlightAlarms(i: Int) {
        if (plates[i]!!.runs == Plate.STARTED && plates[i]!!.checkIfFired()) plateAIT[i]!!.text =
            String.format(resources.getString(R.string.alarming), plateAIT[i]!!.text)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate1(view: View) {
        clickSetButton(0)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate2(view: View) {
        clickSetButton(1)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate3(view: View) {
        clickSetButton(2)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate4(view: View) {
        clickSetButton(3)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate5(view: View) {
        clickSetButton(4)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setPlate6(view: View) {
        clickSetButton(5)
    }

    private fun clickSetButton(plate: Int) {
        if (!alarmSoundBlockSet) {
            val intent = Intent(this, SetPlateActivity::class.java)
            intent.putExtra(SetPlateActivity.PLATE, plate)
            startActivity(intent)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate1(view: View) {
        clickStartButton(0)
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate2(view: View) {
        clickStartButton(1)
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate3(view: View) {
        clickStartButton(2)
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate4(view: View) {
        clickStartButton(3)
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate5(view: View) {
        clickStartButton(4)
    }

    @Suppress("UNUSED_PARAMETER")
    fun startPlate6(view: View) {
        clickStartButton(5)
    }

    /**
     * Serve start buttons and chronometers according plates state
     */
    private fun clickStartButton(plate: Int) {
        val chronometer = chronos[plate]
        val button = startButtons[plate]
        if (plates[plate]!!.isReady && plates[plate]!!.computeSetOff() > 0) {
            startAlarmFromNow(plate)
            chronometer!!.base = plates[plate]!!.baseForChronometer
            //chronometer.format = "00:00:00"
            chronometer.start()
            button!!.setImageDrawable(
                ResourcesCompat.getDrawable(
                    resources,
                    R.drawable.outline_stop_black_36,
                    null
                )
            )
            notification()
        } else if (plates[plate]!!.isStarted) {
            stopAlarm(plate)
            for (alarmed in plates) {
                if (alarmed!!.isStarted && alarmed.checkIfFired()) stopAlarm(alarmed.id)
            }
        } else if (plates[plate]!!.isStopped) {
            chronometer!!.base = SystemClock.elapsedRealtime()
            button!!.setImageDrawable(
                ResourcesCompat.getDrawable(
                    resources,
                    R.drawable.outline_play_arrow_black_36,
                    null
                )
            )
            plates[plate]!!.reset()
            loader!!.savePlate(plate, plates[plate]!!.runs, "")
        }
    }

    private fun stopAlarm(plate: Int) {
        chronos[plate]!!.stop()
        cancelAlarm(plate)
        makeAlarmInfoText(plate)
        alarmSoundServiceStop()
        startButtons[plate]!!.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.baseline_replay_black_36,
                null
            )
        )
        plates[plate]!!.stop()
        if (isAlarm) notification() else cancelNotification()
        plates[plate]!!.last = chronos[plate]!!.text.toString()
        loader!!.savePlate(plate, plates[plate]!!.runs, plates[plate]!!.last)
    }

    private fun startAlarmFromNow(plate: Int) {
        startIntentAndManager(plate, plates[plate]!!.startFromNow())
    }

    private fun startAlarmFromBefore(plate: Int) {
        startIntentAndManager(plate, plates[plate]!!.startFromBefore())
    }

    private fun startIntentAndManager(plate: Int, time: Long) {
        val intent = Intent(this, CTBroadcastReceiver::class.java)
        intent.putExtra(ALARM_OFF_PLATE_NO, plate)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            pIntents[plate] = PendingIntent.getBroadcast(
                this.applicationContext,
                ALARM_UNIQUE_PREFIX + plate,
                intent,
                PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_CANCEL_CURRENT
            )
        }
        (this.getSystemService(ALARM_SERVICE) as AlarmManager)
            .setExact(AlarmManager.RTC_WAKEUP, time, pIntents[plate])
        loader!!.savePlate(
            plate,
            plates[plate]!!.base,
            plates[plate]!!.setOff,
            plates[plate]!!.runs
        )
    }

    private fun cancelAlarm(plate: Int) {
        if (pIntents[plate] != null) {
            (this.getSystemService(ALARM_SERVICE) as AlarmManager)
                .cancel(pIntents[plate])
            pIntents[plate] = null
        } else {
            cancelAlarmManager(plate)
        }
    }

    private fun cancelAlarmManager(plate: Int) {
        (this.getSystemService(ALARM_SERVICE) as AlarmManager)
            .cancel(
                PendingIntent.getBroadcast(
                    this.applicationContext,
                    ALARM_UNIQUE_PREFIX + plate,
                    Intent(this, CTBroadcastReceiver::class.java)
                        .putExtra(ALARM_OFF_PLATE_NO, plate),
                    PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_CANCEL_CURRENT
                )
            )
    }

    private fun onResumeChronometers(i: Int) {
        when (plates[i]!!.runs) {
            Plate.STARTED -> plateIsStartedOnResume(i)
            Plate.STOPPED -> plateIsStoppedOnResume(i)
            Plate.READY -> plateIsReadyOnResume(i)
        }
    }

    private fun plateIsReadyOnResume(i: Int) {
        chronos[i]!!.base = SystemClock.elapsedRealtime()
        startButtons[i]!!.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.outline_play_arrow_black_36,
                null
            )
        )
    }

    private fun plateIsStoppedOnResume(i: Int) {
        chronos[i]!!.text = plates[i]!!.last
        startButtons[i]!!.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.baseline_replay_black_36,
                null
            )
        )
    }

    private fun plateIsStartedOnResume(i: Int) {
        chronos[i]!!.base = plates[i]!!.baseForChronometer
        chronos[i]!!.start()
        startButtons[i]!!.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.outline_stop_black_36,
                null
            )
        )
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
    }

    override fun onStartPressed(plate: Int) {
        //default implementation ignored
    }

    override fun onSetPressed(plate: Int) {
        //default implementation ignored
    }

    private fun makeAlarmInfoText(plate: Int) {
        plateAIT[plate]!!.text = Plate.formatAlarmInfoText(
            plates[plate]!!.hours,
            plates[plate]!!.minutes,
            plates[plate]!!.seconds
        )
    }

    private fun setBackground(i: Int) {
        colorFrame[i]!!.setBackgroundColor(plates[i]!!.color)
    }

    /**
     * initiates UI controls
     */
    private fun initUIControls() {
        initAlarmInfoText()
        initChronometers()
        initStartButtons()
        initBackground()
    }

    /**
     * Obtains references of alarm info texts
     */
    private fun initAlarmInfoText() {
        plateAIT[0] = findViewById<View>(R.id.timeInfo1) as TextView
        plateAIT[1] = findViewById<View>(R.id.timeInfo2) as TextView
        plateAIT[2] = findViewById<View>(R.id.timeInfo3) as TextView
        plateAIT[3] = findViewById<View>(R.id.timeInfo4) as TextView
        plateAIT[4] = findViewById<View>(R.id.timeInfo5) as TextView
        plateAIT[5] = findViewById<View>(R.id.timeInfo6) as TextView
    }

    /**
     * Obtains references of chronometers
     */
    private fun initChronometers() {
        chronos[0] = findViewById<View>(R.id.chronometer1) as Chronometer
        chronos[1] = findViewById<View>(R.id.chronometer2) as Chronometer
        chronos[2] = findViewById<View>(R.id.chronometer3) as Chronometer
        chronos[3] = findViewById<View>(R.id.chronometer4) as Chronometer
        chronos[4] = findViewById<View>(R.id.chronometer5) as Chronometer
        chronos[5] = findViewById<View>(R.id.chronometer6) as Chronometer
    }

    /**
     * Obtains references of start buttons
     */
    private fun initStartButtons() {
        startButtons[0] = findViewById<View>(R.id.button1start) as ImageButton
        startButtons[1] = findViewById<View>(R.id.button2start) as ImageButton
        startButtons[2] = findViewById<View>(R.id.button3start) as ImageButton
        startButtons[3] = findViewById<View>(R.id.button4start) as ImageButton
        startButtons[4] = findViewById<View>(R.id.button5start) as ImageButton
        startButtons[5] = findViewById<View>(R.id.button6start) as ImageButton
    }

    /**
     * Obtains references of background frames
     */
    private fun initBackground() {
        colorFrame[0] = ColorFrame(
            findViewById(R.id.chronometer1),
            findViewById(R.id.timeInfo1),
            findViewById(R.id.buttonLayout1)
        )
        colorFrame[1] = ColorFrame(
            findViewById(R.id.chronometer2),
            findViewById(R.id.timeInfo2),
            findViewById(R.id.buttonLayout2)
        )
        colorFrame[2] = ColorFrame(
            findViewById(R.id.chronometer3),
            findViewById(R.id.timeInfo3),
            findViewById(R.id.buttonLayout3)
        )
        colorFrame[3] = ColorFrame(
            findViewById(R.id.chronometer4),
            findViewById(R.id.timeInfo4),
            findViewById(R.id.buttonLayout4)
        )
        colorFrame[4] = ColorFrame(
            findViewById(R.id.chronometer5),
            findViewById(R.id.timeInfo5),
            findViewById(R.id.buttonLayout5),
            findViewById(R.id.spaceStart)
        )
        colorFrame[5] = ColorFrame(
            findViewById(R.id.chronometer6),
            findViewById(R.id.timeInfo6),
            findViewById(R.id.buttonLayout6),
            findViewById(R.id.spaceEnd)
        )
    }

    companion object {
        const val ALARM_OFF_PLATE_NO = "Alarm_Off_Plate_No"
        private const val NOTIFICATION_ID = 486534515
        private const val ALARM_UNIQUE_PREFIX = 68923402
    }
}